tf2/src/buffer_core.cpp

```diff
@@ -853,15 +860,19 @@ bool BufferCore::canTransform(const std::string& target_frame, const std::string
   if (target_frame == source_frame)
     return true;
 
-  if (warnFrameId("canTransform argument target_frame", target_frame))
+  std::unique_lock<std::mutex> lock(frame_mutex_);
```

Comment:
what are you locking here?
It's preferable locking in scopes to make this clear:

```c++

{
  std::unique_lock<std::mutex> lock(frame_mutex_);
  // ...
}
```

Reply:
I.e. could you validate w/o locking?

Reply:
This line is in the original, it's locking reading of the tf frame tree until the end of the function, for `validateFrameId` and `canTransformNoLock`. It shows up as a new line here because it was moved a few lines up.

Reply:
Is there a risk of validateFrameId to call back canTransform?

Reply:
It's not that validateFrameID has to have a lock held, but that we are pinning the _two_ validation calls to the same version of the tf tree by acquiring the lock. This doesn't seem to be done in all places where two frames are looked up and then have their transforms checked, and the tree walk happens based on a TimePoint, so I don't think it's necessary. My next version only locks for the tree walk itself. Uploading momentarily

Reply:
The new version looks better to me.